### ISSUE TITLE
example of cust el 

### ISSUE THREAD

**Issue by cekvenich** [cekvenich/alp#1]
https://github.com/alpinejs/alpine/issues/272

**Comment by SimoTod** [comment]
I think i need to try it out but I'm not sure i fully get it. 
Alpine is already finding x-data components in your light dom but then it walks down their children to find the other directives. Those element are normal html tag so they don't use the component hook and they aren't found if in the shadow DOM. 
Can you create a working component uaing x-text from the shadow dom when you have time?

**Comment by cekvenich** [comment]
Yes, you should run it.  I think the question was how do you support alpine attributes in CE.
So this example does that. In my mind the next version of Alphine would be based on CE. (also I'd be happy to do all the heavy lifting of the implementation)
x-text would work exactly the same it does not add much, happy to write it, should I do it?

**Comment by cekvenich** [comment]
Also anything else that helps you, let me know.

**Comment by SimoTod** [comment]
Not my call but I think it will be very hard to convince Caleb, he developed Alpine because he wanted something that would work on standard htnl tag. It can be a different framework based on the same concept though. I'll let you know how it goes when I test it.

**Comment by cekvenich** [comment]
CE is standard, built in. Someone will build on it. Happy to support anyone to have a lib that leverages ce

**Comment by SimoTod** [comment]
okay, I played a bit with your example. I'm still not sure if I've got the full picture, yet.

Would it be an alternative way to instantiate component along with the current one or would it be a replacement? At the moment, Alpine uses `document.querySelectorAll('[x-data]')` to instantiate all the components in the normal DOM (including, custom elements).

Even using the attribute observer, we still have to walk down the DOM (and the shadow DOM which would require an Alpine patch) to refresh any directives that we could have, unless you can capture them in a different way, right?

If you want to work on a slightly more sophisticated example, I'll be happy to have a look but I need a bit more context on what you want to achieve. 
Whether you want to only support CE or also alpine component inside the shadow DOM of a CE.
What benefit you will get from the approach, etc.

A good example would be a modal where the slot is the modal content, hidden by default and it shows when it receives a specific custom event.

**Comment by cekvenich** [comment]
Cool that you checked it. 

As to having to refresh directives | walk the tree: that is not needed in this CE approach. Any attribute set in DOM by page can also be set by page in .js. And if one CE wanted to send a message to all or another CE it would be done via Events. As you pointed out via Custom Event (or I use a EventBus). So any refresh or walk the tree = Events.

As to what I want to achieve: I'm open to a few things, including collaborating on what that idea is. I 
I love that Alpine is DOM centric relative to other popular libs. Also, clearly CE are the future. 
I hope that it is worth exploring ideas on how to combine these two things. What are some possible benefits? Likely the lib would be smaller, faster, easier to add plugins, and more standard due to it being native. Or maybe not, but at least worth experimenting. Maybe other benefits appear or other weaknesses appear. 
At initial glance, it appears that Alpine can support CE or even Alpine can be implemented via CE. 
I can for sure take all the directives and magic and do it via CE. Is it better or worse, I have no foresight other than CE will dominate and replace some popular libs. I do believe that the DOM approach is superior to .js approaches, and CE helps hide .js, so CE would only sharpen that.

So yes, I can write a modal that waits for a custom event to show, let me do that. 
The 'real' implementation would use composition I think, but for now, I'll stick with inheritance as faster to type. I'll ping you when I have it.

And even if there are better / more realistic business use cases.

**Comment by cekvenich** [comment]
@SimoTod  done! Works in Chrome. 
Does that help? I think cleaner and less code.

**Comment by cekvenich** [comment]
@SimoTod Do you plan to look at this? :-)

**Comment by SimoTod** [comment]
Hi @cekvenich I got the general gist but I'm still a bit unsure how this implementation would relate to Alpine. The second example doesn't use any of the Alpine stuff, it looks like a normal custom component to implement a modal.

I don't want to put you off, I think custom element are a really interesting topic and they should be used in web development. You can definitely work on your own framework as well.

However, I'm not sure if this kind of implementation (which, if I've understood correctly, would move all the code to an abstract custom element) would be considered by Caleb.

I'm not an Alpine maintainer (I've just sent a few PRs where I saw I could help), Caleb is the only person maintaining Alpine so maybe I'm wrong and you can try to ask him.

**Comment by cekvenich** [comment]
Can you connect him with me? I'm vic(at)eml.cc.

That would be idea, to write a Base CE that mimics Alpine grammar. I'm happy to write it, but only if someone helps me do some product management.
